vsr/docker/launch.py

```diff
@@ -0,0 +1,161 @@
+
+
+class VSR_vm(vrnetlab.VM):
+    def __init__(self, username, password):
+        for e in os.listdir("/"):
+            if re.search(".qcow2$", e):
+                disk_image = "/" + e
+        super(VSR_vm, self).__init__(username, password, disk_image=disk_image, ram=1024)
+        self.qemu_args.extend(["-boot", "n", "-monitor", "tcp:0.0.0.0:5001,server,nowait"])
+        self.num_nics = 4
+
+    def bootstrap_spin(self):
+        """ This function should be called periodically to do work.
+        """
+
+        if self.spins > 300:
+            # too many spins with no result ->  give up
+            self.stop()
+            self.start()
+            return
+
+        (ridx, match, res) = self.tn.expect([b"Performing automatic"], 1)
+        if match: # got a match!
+            if ridx == 0: # login
+                self.logger.debug("VM started")
+
+                self.logger.debug("Connecting to QEMU Monitor")
+                self.tn = telnetlib.Telnet("127.0.0.1", 5001 + self.num)
+                self.wait_write("", wait=")")
+
+                # To allow access to aux0 serial console
+                self.logger.debug("Writing to QEMU Monitor")
+                with open("qemu.txt", "r+") as file:
+                    for line in file.readlines():
+                        self.wait_write(line, wait=")")
+                        # self.logger.debug("Wrote line:" + line)
+                        time.sleep(0.1)
+                file.close()
+
+                self.wait_write("", wait=")")
+                self.logger.debug("Done writing to QEMU Monitor")
+
+                self.logger.debug("Switching to line aux0")
+                self.tn.close()
+                # time.sleep(120)
+                self.tn = telnetlib.Telnet("127.0.0.1", 5000 + self.num)
+
+                # run main config!
+                self.bootstrap_config()
+                # close telnet connection
+                self.tn.close()
+                # startup time?
+                startup_time = datetime.datetime.now() - self.start_time
+                self.logger.info("Startup complete in: %s" % startup_time)
+                # mark as running
+                self.running = True
+                return
+
+        # no match, if we saw some output from the router it's probably
+        # booting, so let's give it some more time
+        if res != b'':
+            self.logger.trace("OUTPUT: %s" % res.decode())
+            # reset spins if we saw some output
+            self.spins = 0
+
+        self.spins += 1
+
+        return
+
+    def bootstrap_config(self):
+        """ Do the actual bootstrap config
+        """
+        self.logger.info("applying bootstrap configuration")
+        self.wait_write("\x0D", None)
+        self.wait_write("\x0D", "<HPE>")
+        self.logger.debug("Entering system view")
+        self.wait_write("system-view", "<HPE>")
+        time.sleep(0.2)
+        self.wait_write("ssh server enable", None)
```

Comment:
Is there really no prompt here? Your previous call of `self.wait_write` waits for a prompt that looks like `<HPE>`. Do you not have that prompt here?

It is much better to wait for a prompt rather than use `time.sleep` to pace output.